apps/meteor/client/views/room/providers/ComposerPopupProvider.tsx

```diff
@@ -64,14 +67,16 @@ const ComposerPopupProvider = ({ children, room }: ComposerPopupProviderProps) =
 					const filterRegex = filter && new RegExp(escapeRegExp(filter), 'i');
 					const items: ComposerBoxPopupUserProps[] = [];
 
-					const users = Array.from(usersFromRoomMessages.values())
-						.filter((u) => !!u.ts && (filterRegex ? u.username.match(filterRegex) || u.name?.match(filterRegex) : true))
+					const roomMessageUsers = queryClient.getQueryData<RoomMessageUser[]>(roomMessageUsersQueryKeys.all(rid, uid)) ?? [];
+
+					roomMessageUsers
 						.sort((a, b) => b.ts.getTime() - a.ts.getTime())
-						.slice(0, suggestionsCount ?? 5)
-						.map((u) => {
-							u.suggestion = true;
-							return u;
-						});
+						.filter((u) => {
+							if (!filterRegex) return true;
+							return filterRegex.test(u.username) || (u.name && filterRegex.test(u.name));
+						})
+						.slice(0, suggestionsCount ?? 5);
```

Comment:
This is more of an exercise to improve this. Each operation here will loop through the array list some times. Can you think of a better order for these operations so we can do less work on the array?